### PR TITLE
Remove std::map from scatter_values in nonmatching interpolation

### LINKED ISSUE #2419: Remove `std::map` from non-matching mesh interpolation
See e.g. https://github.com/FEniCS/dolfinx/pull/2245/files#diff-6208a9ad6f57063c470b05140af6f96ed580abdce54401eab6b0f89df2344ddaR106

### PR DESCRIPTION
**Pull request by jorgensd**
Resolves #2419.

### CODE REVIEW

**Comment by jorgensd** [comment]
@massimiliano-leoni do you mind testing this on your huge secret meshes and see that we don't get a regression?
 I can't see how it would regress, as we add some optimizations here to one of the `rank_to_neighbor` maps.

File: cpp/dolfinx/fem/interpolate.h [hunk 1]
```diff
@@ -194,9 +194,10 @@ void scatter_values(MPI_Comm comm, std::span<const std::int32_t> src_ranks,
   std::vector<std::int32_t> recv_offsets(in_ranks.size() + 1, 0);
   {
     // Build map from parent to neighborhood communicator ranks
-    std::map<std::int32_t, std::int32_t> rank_to_neighbor;
+    std::vector<std::pair<std::int32_t, std::int32_t>> rank_to_neighbor;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:197]: Add `rank_to_neighbor.reserve(in_ranks.size());`

File: cpp/dolfinx/fem/interpolate.h [hunk 2]
```diff
@@ -233,14 +242,26 @@ void scatter_values(MPI_Comm comm, std::span<const std::int32_t> src_ranks,
   send_sizes.reserve(1);
   {
     // Compute map from parent mpi rank to neigbor rank for outgoing data
-    std::map<std::int32_t, std::int32_t> rank_to_neighbor;
+    // `out_ranks` are already sorted so ranK_to_neighbor is sorted
+    std::vector<std::pair<std::int32_t, std::int32_t>> rank_to_neighbor;
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:246]: `rank_to_neighbor.reserve(out_ranks.size());`

File: cpp/dolfinx/fem/interpolate.h [hunk 3]
```diff
@@ -233,14 +242,26 @@ void scatter_values(MPI_Comm comm, std::span<const std::int32_t> src_ranks,
...
+      rank_to_neighbor.push_back({out_ranks[i], i});
 
     // Compute send sizes
+    // As `src_ranks` are sorted we can move start in search forward
+    auto start = rank_to_neighbor.begin();
     std::ranges::for_each(
-        src_ranks, [&rank_to_neighbor, &send_sizes, block_size](auto rank)
-        { send_sizes[rank_to_neighbor[rank]] += block_size; });
+        src_ranks,
+        [&rank_to_neighbor, &send_sizes, &block_size, &start](auto rank)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:255]:
```suggestion
        [&rank_to_neighbor, &send_sizes, block_size, &start](auto rank)
```

File: cpp/dolfinx/fem/interpolate.h [hunk 4]
```diff
@@ -233,14 +242,26 @@ void scatter_values(MPI_Comm comm, std::span<const std::int32_t> src_ranks,
   send_sizes.reserve(1);
   {
     // Compute map from parent mpi rank to neigbor rank for outgoing data
-    std::map<std::int32_t, std::int32_t> rank_to_neighbor;
+    // `out_ranks` are already sorted so ranK_to_neighbor is sorted
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:245]:
```suggestion
    // `out_ranks` is sorted, so rank_to_neighbor will be sorted too
```

File: cpp/dolfinx/fem/interpolate.h [hunk 5]
```diff
@@ -233,14 +242,26 @@ void scatter_values(MPI_Comm comm, std::span<const std::int32_t> src_ranks,
...
+    std::vector<std::pair<std::int32_t, std::int32_t>> rank_to_neighbor;
     for (std::size_t i = 0; i < out_ranks.size(); i++)
-      rank_to_neighbor[out_ranks[i]] = i;
+      rank_to_neighbor.push_back({out_ranks[i], i});
 
     // Compute send sizes
+    // As `src_ranks` are sorted we can move start in search forward
```
Comment (garth-wells) [inline cpp/dolfinx/fem/interpolate.h:251]:
```suggestion
    // As `src_ranks` is sorted, we can move 'start' in search forward
```

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/interpolate.h (hunk 1 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/interpolate.h:197]:
```suggestion
    std::vector<std::pair<std::int32_t, std::int32_t>> rank_to_neighbor;
    rank_to_neighbor.reserve(in_ranks.size());
```

File: cpp/dolfinx/fem/interpolate.h (hunk 2 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/interpolate.h:246]:
```suggestion
    std::vector<std::pair<std::int32_t, std::int32_t>> rank_to_neighbor;
    rank_to_neighbor.reserve(out_ranks.size());
```

**Review by garth-wells — APPROVED** [review]

**Pull request merged**